From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Sun, 19 Jan 2025 12:53:05 +0800
Subject: [PATCH 41/50] AOSCOS: drm: loongson: add ls7a2000_support module
 parameter

The loongson DRM driver does not implement userspace acceleration,
resulting in a lack of OpenGL/3D acceleration support. Loongson Technology
supplies a LoongGPU driver package which ships with a more complete
feature set. Users of 7A2000-based graphics are likely to get a better
experience out of this driver package instead.

Disable loongson driver probing by default for LS7A2000/2K2000 devices.

For those who feel adventurous, use the `loongson.ls7a2000_support=1'
parameter to enable this driver.

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

Link: https://t.me/c/1109254909/723685
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

[ Rong Bao: Removed dependency on a prior patch, "AOSCOS: drm: loongson:
  add ls7a1000_support module parameter" ]
Signed-off-by: Rong Bao <rong.bao@csmantle.top>
---
 drivers/gpu/drm/loongson/loongson_module.c | 4 ++++
 drivers/gpu/drm/loongson/loongson_module.h | 1 +
 drivers/gpu/drm/loongson/lsdc_drv.c        | 8 ++++++++
 3 files changed, 13 insertions(+)

diff --git a/drivers/gpu/drm/loongson/loongson_module.c b/drivers/gpu/drm/loongson/loongson_module.c
index d2a51bd395f6c9153b155515802bee7af025a885..61dc82d161f30f5aef8b00c7b5635fcf313f33c2 100644
--- a/drivers/gpu/drm/loongson/loongson_module.c
+++ b/drivers/gpu/drm/loongson/loongson_module.c
@@ -17,6 +17,10 @@ int loongson_vblank = 1;
 MODULE_PARM_DESC(vblank, "Disable/Enable hw vblank support");
 module_param_named(vblank, loongson_vblank, int, 0400);
 
+int loongson_ls7a2000_support = 0;
+MODULE_PARM_DESC(ls7a2000_support, "7A2000 support (1 = enabled, 0 = disabled (default))");
+module_param_named(ls7a2000_support, loongson_ls7a2000_support, int, 0400);
+
 static int __init loongson_module_init(void)
 {
 	if (!loongson_modeset || video_firmware_drivers_only())
diff --git a/drivers/gpu/drm/loongson/loongson_module.h b/drivers/gpu/drm/loongson/loongson_module.h
index 931c17521bf017d65422cea1a2e18f9ece1bfb6e..fbe2cd8f90ff7721d93ee6e04b1db545fa4b992b 100644
--- a/drivers/gpu/drm/loongson/loongson_module.h
+++ b/drivers/gpu/drm/loongson/loongson_module.h
@@ -7,6 +7,7 @@
 #define __LOONGSON_MODULE_H__
 
 extern int loongson_vblank;
+extern int loongson_ls7a2000_support;
 extern struct pci_driver lsdc_pci_driver;
 
 #endif
diff --git a/drivers/gpu/drm/loongson/lsdc_drv.c b/drivers/gpu/drm/loongson/lsdc_drv.c
index 34405073c4d48e669f390e2bbfb3084def211994..08587d92fb9d905eb0c507f21139cb431d0a6f81 100644
--- a/drivers/gpu/drm/loongson/lsdc_drv.c
+++ b/drivers/gpu/drm/loongson/lsdc_drv.c
@@ -267,6 +267,14 @@ static int lsdc_pci_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
 	int ret;
 
 	descp = lsdc_device_probe(pdev, ent->driver_data);
+
+	if (to_loongson_gfx(descp)->chip_id == CHIP_LS7A2000 &&
+	    !loongson_ls7a2000_support) {
+		dev_info(&pdev->dev,
+			"LS7A2000 support disabled or not enabled via module parameter\n");
+		return -ENODEV;
+	}
+
 	if (IS_ERR_OR_NULL(descp))
 		return -ENODEV;
 
-- 
2.55.0

